pkg/nsx/services/vpc/vpc.go

```diff
@@ -451,10 +453,45 @@ func (s *VPCService) GetAVISubnetInfo(vpc model.Vpc) (string, string, error) {
 	return path, cidr, nil
 }
 
-func (s *VPCService) CreateorUpdateVPC(obj *v1alpha1.VPC) (*model.Vpc, *common.VPCNetworkConfigInfo, error) {
+func (s *VPCService) CreateOrUpdateVPC(obj *v1alpha1.NetworkInfo) (*model.Vpc, *common.VPCNetworkConfigInfo, error) {
 	// check from VPC store if vpc already exist
+	ns := obj.Namespace
 	updateVpc := false
-	existingVPC := s.VpcStore.GetVPCsByNamespace(obj.Namespace)
+	nsObj := &v1.Namespace{}
+	// get name obj
+	if err := s.Client.Get(ctx, types.NamespacedName{Name: obj.Namespace}, nsObj); err != nil {
+		log.Error(err, "unable to fetch namespace", "name", obj.Namespace)
+		return nil, nil, err
+	}
+	annotations := nsObj.GetAnnotations()
+	vpcName, nameExist := annotations[common.AnnotationVPCName]
+	if nameExist {
+		log.Info("read ns annotation networkInfoName", "VPCNAME", vpcName)
+		res := strings.Split(vpcName, "/")
+		// The format should be namespace/vpc_name
+		if len(res) != 2 {
+			message := fmt.Sprintf("incorrect networkInfoName annotation %s for namespace %s", vpcName, obj.Namespace)
+			return nil, nil, errors.New(message)
+		}
+		if ns != res[0] {
+			log.Info("namespace is using shared vpc, with vpc name anno", "VPCNAME", vpcName, "Namespace", ns)
+			ns = res[0]
+			networkInfos := &v1alpha1.NetworkInfoList{}
+			s.Client.List(context.Background(), networkInfos, client.InNamespace(ns))
+			if len(networkInfos.Items) > 0 {
+				// if there is already one networkInfo, clear it's vpc and update
+				sharedNetworkInfoCR := &v1alpha1.NetworkInfo{}
+				err := s.Client.Get(context.Background(), client.ObjectKey{Namespace: ns, Name: networkInfos.Items[0].Name}, sharedNetworkInfoCR)
+				if err != nil {
+					return nil, nil, err
+				}
+				// transfer the obj to the shared obj
+				obj = sharedNetworkInfoCR
+			}
+		}
+	}
+
```

Comment:
For shared VPC case, the below logic would not be needed. We can just "return" here.

Reply:
For shared, the below is for updating case. For example, vmware-system is shareing kube-system's vpc, any namespace arrives first, it will update vpc if vpcnetworkconfig changes.

Reply:
OK, fixed as discussed.